<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

<b>Инструкция по выполнению проекта:</b>
<br>
- Загрузите и изучите данные.
<br>

- Ответьте на вопрос и обоснуйте решение.
<br>
Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)<br>
a. Изменится. Приведите примеры матриц.<br>
b. Не изменится. Укажите, как связаны параметры линейной регрессии в исходной задаче и в преобразованной.
<br>

- Предложите алгоритм преобразования данных для решения задачи. Обоснуйте, почему качество линейной регрессии не поменяется.<br>

- Запрограммируйте этот алгоритм, применив матричные операции. Проверьте, что качество линейной регрессии из sklearn не отличается до и после преобразования. Примените метрику R2.
<br>

<b>Укажем описание признаков:</b>

<b>Признаки</b>
- пол застрахованного
- возраст застрахованного
- зарплата застрахованного
- количество членов его семьи

<b>Целевой признак</b>
- количество страховых выплат клиенту за последние 5 лет

## Загрузка данных

In [1]:
#подключим необходимые для работы библиотеки

import pandas as pd
import numpy as np
import warnings
warnings.simplefilter("ignore")
pd.options.display.float_format = '{:,.3f}'.format

from sklearn.linear_model import LinearRegression 
from sklearn.model_selection import train_test_split 

from sklearn.metrics import r2_score

In [2]:
#прочитаем исходный датафрейм

try:
    data = pd.read_csv('C:/Users/jnybfuvybtfgl kfolbl/mypythonworks/yandex/insurance.csv', sep=',') 
    
except:
    data = pd.read_csv('/datasets/insurance.csv', sep=',')

FileNotFoundError: [Errno 2] No such file or directory: '/datasets/insurance.csv'

In [3]:
# изучим основную информацию о датафрейме

print('Первые 10 строк датафрейма:')
display(data.head(10))
print('Основная информация о датафрейме:')
display(data.info())
print('Размер датасета:')
display(data.shape)
print('Доля пропущенных данных в датафрейме:')
display(data.isna().mean())
print('Количество дубликатов: ')
display(data.duplicated().sum())

Первые 10 строк датафрейма:


Пол  Возраст   Зарплата  Члены семьи  Страховые выплаты
0    1   41.000 49,600.000            1                  0
1    0   46.000 38,000.000            1                  1
2    0   29.000 21,000.000            0                  0
3    0   21.000 41,700.000            2                  0
4    1   28.000 26,100.000            0                  0
5    1   43.000 41,000.000            2                  1
6    1   39.000 39,700.000            2                  0
7    1   25.000 38,600.000            4                  0
8    1   36.000 49,700.000            1                  0
9    1   32.000 51,700.000            1                  0

Основная информация о датафрейме:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


None

Размер датасета:


(5000, 5)

Доля пропущенных данных в датафрейме:


Пол                 0.000
Возраст             0.000
Зарплата            0.000
Члены семьи         0.000
Страховые выплаты   0.000
dtype: float64

Количество дубликатов: 


153

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
Изучив исходные данные датафрейма можно сказать, что данные представлены в полном объеме. Типы данных корректны, но дополнительно переведем тип данных у столбцов 'Пол' и 'Возраст' в целочисленный, т.к. у них отсутствуют данные после запятой. 
</div>

In [4]:
#преобразуем типы данных

data['Пол']=data['Пол'].astype('int')
data['Возраст']=data['Возраст'].astype('int')

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
Также у нас обнаружилось 153 дубликата. Но т.к. у нас нет уникального ID клиента, то вполне вероятно, при большой клиентской базе совпадение определенных признаков. Так что оставляем данные дубликаты в нашем датафрейме.
</div>

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Вопрос:** Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии?

**Ответ:** не изменится.<br>

Заменим матрицу $X$ на матрицу $Z$:

$$
Z = XP \qquad (2.1)
$$

где $P$ - обратимая матрица с некими значениями, на которую может быть умножена матрица $X$. 

**Обоснование:** <br>

Заменим $X$ на $Z$ и вычислим, чему будет равно предсказание и вектор весов.

$$
a_1 = Zw_1 \qquad (2.2)
$$
$$
w_1 = (Z^T Z)^{-1} Z^T y \qquad (2.3)
$$

Подставим в уравнение 2.2 правую часть уравнения 2.3 и получим следущее:

$$
w_1 = Z (Z^T Z)^{-1} Z^T y \qquad (2.4)
$$

Преобразуем выражение для $w1$:  
$$
w_1 = ((XP)^T(XP))^{-1}(XP)^Ty
$$

$$
w_1 = P^{-1}((XP)^{T}X)^{-1}(XP)^Ty
$$

$$
w_1 = P^{-1}((XP)^{T}X)^{-1}P^TX^Ty
$$

$$
w_1 = P^{-1}(P^TX^TX)^{-1}P^TX^Ty
$$ 

$$
w_1 = P^{-1}(X^TX)^{-1}(P^T)^{-1}P^TX^Ty
$$

Сократим

$(P^{T})^{-1}$ и $P^T$

получается
$$
w_1 = P^{-1}(X^TX)^{-1}  X^Ty
$$
Произведем замену:
$$
w_1 = (X^TX)^{-1}X^Ty = P^{-1}w
$$ 
Тогда:
$$
w_1 = P^{-1}w 
$$
Подставим для $X$ и $w1$ в формулу для расчета $a1$:
$$
a_1 =Xw_1 = XPP^{-1}w = Xw = a
$$

Выражения равны, ч.т.д.

## Алгоритм преобразования

**Алгоритм**

Для защиты информации на этапе поиска коэффициентов регрессии и предсказаний регрессии будем умножать матрицу признаков на обратимую матрицу $Y$, которая будет генерироваться случайным образом.

Этапы алгоритма:
1. Составление матрицы $Y$.
2. Проверка матрицы на обратимость. Вычисление детерминанта матрицы $Y$.
3. Получение матрицы преобразованных признаков $Z = XY$.
4. Применение алгоритма на преобразованных признаках $Z$.

**Обоснование**

Матрица $Y$ должна иметь необходимую размерность $(nxn)$, где n - количество признаков для регрессии.
Таким образом матрица $Z$ будет иметь ту же размерность, что и матрица $X$. Обратная матрица $Y$ существует только для квадратных невырожденных матриц (определитель которых не равен нулю).

Пример:

$
X = \begin{pmatrix}
1 & 2 \\
2 & 3 \\
4 & 5  
\end{pmatrix}
\qquad 
Y = \begin{pmatrix}
1 & 0 \\
2 & 3 \\  
\end{pmatrix}
\qquad det  Y = 3 $

Найдем значние Z:

$
Z = \begin{pmatrix}
1 & 2 \\
2 & 3 \\
4 & 5  
\end{pmatrix} \begin{pmatrix}
1 & 0 \\
2 & 3 \\  
\end{pmatrix} = \begin{pmatrix}
1*1+2*2 & 1*0+2*3  \\
2*1+3*2 & 2*0+3*3  \\
4*1+5*2 & 4*0+5*3   
\end{pmatrix} = \begin{pmatrix}
5 & 6 \\
8 & 9 \\
14 & 15  
\end{pmatrix}$

После этого добавляем нулевой столбец и вводим данные в линейную регрессию.

## Проверка алгоритма

<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
Для начала проверим качество модели линейной регрессии на данных без преобразования.
</div>

In [5]:
#для начала отделим обучающую выборку от тестовой

features = data.drop('Страховые выплаты',axis=1)
target = data['Страховые выплаты']

features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.25, 
                                                                              random_state=12345)
#далее узнаем размеры выборок

print('Размер обучающей выборки равен: ', features_train.shape)
print('Размер тестовой выборки равен: ', features_test.shape)

Размер обучающей выборки равен:  (3750, 4)
Размер тестовой выборки равен:  (1250, 4)


In [6]:
#обучим модель и узнаем качество модели

model = LinearRegression()
model.fit(features_train, target_train)

predictions = model.predict(features_test)

r2_score_init = r2_score(target_test, predictions)
print("Качество модели на начальных данных: {:.3f}".format(r2_score_init))

Качество модели на начальных данных: 0.435


<div class="alert alert-info">
<b>Комментарий студента:</b>
<br>
Далее произведем преобразование данных и исследуем качество модели линейной регрессии.
</div>

In [7]:
#cоздадим случайную обратимую матрицу

random_matrix = np.random.normal(size=(features_train.shape[1], features_train.shape[1]))

det = np.linalg.det(random_matrix)

print("Детерминант случайной матрицы равен: {:.3f}, значит матрица обратима.".format(det))

Детерминант случайной матрицы равен: 0.155, значит матрица обратима.


In [8]:
#произведем умножение случайной обратимой матрицы на матрицу признаков

features_train_transform = np.dot(features_train, random_matrix)
features_test_transform = np.dot(features_test, random_matrix)

#матрица для получения исходной матрицы
random_matrix_inv = np.linalg.inv(random_matrix)
 
# создадим класс
class Regression_transform:
    
    def fit(self, features_train, target_train):
       
        X = np.concatenate((np.ones((features_train.shape[0], 1)), features_train), axis=1)
        y = target_train
        w = np.linalg.inv(np.dot(X.T, X)).dot(X.T).dot(y)
        self.w = w[1:]
        self.w0 = w[0]   
        
    def predict(self, features_test):
        return features_test.dot(self.w) + self.w0

#инициализируем модель  
model_transform = Regression_transform()

#обучим модель на преобразованных данных
model_transform.fit(features_train_transform, target_train)
predictions_test = model_transform.predict(features_test_transform)

#узнаем качество модели на преобразованных данных
r2_transform = r2_score(target_test, predictions_test)

print("Качество модели на преобразованных данных: {:.3f}".format(r2_transform))

Качество модели на преобразованных данных: 0.435


In [9]:
#для примера получим исходную тестовую выборку до преобразования

init_matrix_test = np.dot(features_test_transform, random_matrix_inv)
init_matrix_test

array([[-2.35156100e-12,  3.30000000e+01,  3.90000000e+04,
         4.00000000e+00],
       [-9.66568103e-12,  5.00000000e+01,  4.31000000e+04,
         2.00000000e+00],
       [ 1.00000000e+00,  3.90000000e+01,  4.21000000e+04,
        -2.18651731e-12],
       ...,
       [ 1.00000000e+00,  3.50000000e+01,  4.23000000e+04,
        -6.91044215e-12],
       [ 1.00000000e+00,  3.60000000e+01,  4.13000000e+04,
        -2.62947020e-12],
       [-5.52878594e-12,  4.90000000e+01,  3.11000000e+04,
         1.00000000e+00]])

In [10]:
#преобразуем целевой признак в датафрейм

target_test_init = pd.DataFrame(target_test)

In [11]:
#преобразуем признаки в датафрейм

features_test_init = pd.DataFrame(init_matrix_test, index=target_test_init.index, columns=features_test.columns).join(target_test_init)

In [12]:
features_test_init

Пол  Возраст   Зарплата  Члены семьи  Страховые выплаты
3183 -0.000   33.000 39,000.000        4.000                  0
1071 -0.000   50.000 43,100.000        2.000                  2
2640  1.000   39.000 42,100.000       -0.000                  0
2282  0.000   20.000 34,800.000        0.000                  0
1595  0.000   41.000 40,000.000        4.000                  0
...     ...      ...        ...          ...                ...
982   1.000   51.000 29,000.000        2.000                  2
3820  1.000   33.000 57,900.000        3.000                  0
3595  1.000   35.000 42,300.000       -0.000                  0
3513  1.000   36.000 41,300.000       -0.000                  0
1205 -0.000   49.000 31,100.000        1.000                  2

[1250 rows x 5 columns]

<div class="alert alert-info">
<b>ВЫВОД:</b>
<br>
В данном проекте было проверено и доказано суждение о том, что если признаки умножить на обратимую матрицу, то ее качество не изменится. Также была проведена работа по дешифровке преобразованных данных.
</div>